lib/next_ls/helpers/ast_helpers.ex

```diff
@@ -204,19 +204,68 @@ defmodule NextLS.ASTHelpers do
     end
   end
 
-  def find_cursor(ast) do
-    with nil <-
-           ast
-           |> Zipper.zip()
-           |> Zipper.find(fn
-             {:@, _, [{:__cursor__, _, []}]} -> true
-             {:__cursor__, _, _} -> true
-             {{:., _, [_, :__cursor__]}, _, _} -> true
-             _ -> false
-           end) do
-      {:error, :not_found}
+  def find_cursor(ast, position) do
+    ast = {:__block__, [], [ast]}
+
+    {tree, cursor_tree} =
+      ast
+      |> Zipper.zip()
+      |> Zipper.traverse(nil, fn tree, acc ->
+        node = Zipper.node(tree)
+        node_range = Sourceror.Range.get_range(node)
+        is_literal = Macro.quoted_literal?(node)
+
+        is_inside =
+          with nil <- node_range, is_literal do
+            false
+          else
+            _ -> sourceror_inside?(node_range, position)
+          end
+
+        cond do
+          is_literal ->
+            {tree, acc}
+
+          is_inside ->
+            {tree, tree}
+
+          true ->
+            {tree, acc}
+        end
+      end)
+
+    if cursor_tree do
+      left = Zipper.left(cursor_tree)
+      up = Zipper.up(cursor_tree)
+
+      cond_result =
+        cond do
+          up && match?({:"::", _, _}, Zipper.node(up)) ->
+            up
+            |> Zipper.up()
+            |> Zipper.update(fn n ->
+              {:__block__, [],
+               [
+                 {:__cursor__, [], []},
+                 n
+               ]}
+            end)
+
+          up && match?({:<-, _, _}, Zipper.node(up)) ->
+            up |> Zipper.insert_left({:__cursor__, [], []}) |> Zipper.down() |> Zipper.rightmost()
+
+          left && Zipper.node(left) == :do ->
+            Zipper.update(cursor_tree, fn n ->
+              {:__block__, [], [n, {:__cursor__, [], []}]}
+            end)
+
+          true ->
+            Zipper.insert_right(cursor_tree, {:__cursor__, [], []})
+        end
+
+      Zipper.top(cond_result)
```

Comment:
Maybe you can return cond_result as well, wouldn't that be the zipper being at the cursor which you find in the `with` clause in the build step.